launchpad/src/launchpad.rs

```diff
@@ -3,100 +3,122 @@
 elrond_wasm::imports!();
 elrond_wasm::derive_imports!();
 
-use elrond_wasm::elrond_codec::TopEncode;
-
 mod setup;
 
-mod launch_stage;
-use launch_stage::*;
-
 mod ongoing_operation;
 use ongoing_operation::*;
 
 mod random;
 use random::Random;
 
-mod ticket_status;
-use ticket_status::TicketStatus;
+const FIRST_TICKET_ID: usize = 1;
+
+type TicketStatus = bool;
+const WINNING_TICKET: TicketStatus = true;
+
+#[derive(TopEncode, TopDecode)]
+pub struct TicketRange {
+    pub first_id: usize,
+    pub last_id: usize,
+}
 
-const VEC_MAPPER_START_INDEX: usize = 1;
-const FIRST_GENERATION: u8 = 1;
+#[derive(TopEncode, TopDecode)]
+pub struct TicketBatch {
+    pub address: Address,
+    pub nr_tickets: usize,
+}
 
 #[elrond_wasm::derive::contract]
 pub trait Launchpad: setup::SetupModule + ongoing_operation::OngoingOperationModule {
-    // endpoints - owner-only
-
     #[only_owner]
     #[endpoint(claimTicketPayment)]
     fn claim_ticket_payment(&self) -> SCResult<()> {
-        let ticket_payment_token = self.ticket_payment_token().get();
-        let sc_balance = self.blockchain().get_sc_balance(&ticket_payment_token, 0);
+        self.require_claim_period()?;
+
         let owner = self.blockchain().get_caller();
 
-        if sc_balance > 0 {
-            self.send()
-                .direct(&owner, &ticket_payment_token, 0, &sc_balance, &[]);
+        let ticket_payment_claimable_amount = self.claimable_ticket_payment().get();
+        if ticket_payment_claimable_amount > 0 {
+            let ticket_payment_token = self.ticket_payment_token().get();
+
+            self.claimable_ticket_payment().clear();
+            self.send().direct(
+                &owner,
+                &ticket_payment_token,
+                0,
+                &ticket_payment_claimable_amount,
+                &[],
+            );
+        }
+
+        // this only happens when too many users are blacklisted/don't confirm enough tickets
+        let launchpad_token_id = self.launchpad_token_id().get();
+        let nr_winning_tickets = self.nr_winning_tickets().get();
+        let nr_claimed_tickets = self.total_claimed_tickets().get();
+        let nr_tickets_left_to_claim = nr_winning_tickets - nr_claimed_tickets;
+
+        let launchpad_tokens_per_winning_ticket = self.launchpad_tokens_per_winning_ticket().get();
+        let total_lauchpad_tokens_needed =
+            Self::BigUint::from(nr_tickets_left_to_claim) * launchpad_tokens_per_winning_ticket;
+        let sc_balance = self.blockchain().get_sc_balance(&launchpad_token_id, 0);
+
+        if sc_balance > total_lauchpad_tokens_needed {
+            let leftover_launchpad_tokens = sc_balance - total_lauchpad_tokens_needed;
+            self.send().direct(
+                &owner,
+                &launchpad_token_id,
+                0,
+                &leftover_launchpad_tokens,
+                &[],
+            );
```

Comment:
Funds will most likely be lost here because the `nr_winning_tickets` keeps track of the real number of tickets, but sc_balance keeps track of the rolling balance - the total ammount minus whatever the users have already claimed through `claimLaunchpadTokens`.
The only case where this returns the correct number of launchpad tokens is when the owner calls `claimTicketPayment` before anyone else calls `claimLaunchpadTokens`.
My suggestion would be to move this code in the `OperationCompletionStatus::Completed` stage of the `filterTickets` endpoint.